### PR TITLE
VAD 샘플링하고 실제로 학습하기

### PR DESCRIPTION
**Pull request by daniel03c1**
랜덤으로 config를 생성하고 VAD를 학습하고 테스트하는 과정까지 작성하였습니다.
1. vad_architecture라는 이름으로 model를 만들었고, 그에 맞게 complexity, sampler를 작성하였습니다.
    구조 관련해서 더 좋은 아이디어나 의견 있으시면 적극 반영하겠습니다.
2. 모델 관련 complexity는 따로 model_complexity로 분리시켰습니다.
3. vad에 맞게 dataset 생성을 두가지로 하였습니다. 첫번째는 file name들을 받아 dataset을 넘겨주는 형태, 
    두번째는 미리 전처리된 features, labels를 받아 넘겨서 dataset을 넘기는 형태 두가지입니다.
    공통사항: 매 wav가 길이가 다르기 때문에 generator로 미리 넘겨서 cache한 이후에, window를 적용하는
    함수를 통해 매번 랜덤으로 window를 적용하는 방식입니다. 
    파일 단위로 generator가 생성하도록 한 이후에 cache하는 이유는, generator에서 직접 윈도우를 
    생성하게 하면, 캐시기능을 사용하지 못해 매우 느려집니다.
4. VAD의 search space
    DNN, bDNN 모두 대략 flops가 55만 정도 되어, 50만에서 60만 사이로 설정하였습니다.
    search space 관련해서는 여러분들의 의견이 많이 필요합니다. 일단 작은 flops에서 찾도록 하기 위해 depth의
    최대한도를 3으로 하여 sampling이 진행되도록 하였습니다.
5. VAD 학습 방식
    loss가 기존 코드는 MSE로 되어 있는데 binary crossentropy 등이 좋지 않을까 생각이 들어 바꾸어 진행하였습니다.
    augmentation 등이 빠져있는데, 기존 코드에서 아무 augmentation을 사용하지 않아서 그대로 따랐습니다.
    augmentation이 추가로 정해지면 추가하는 것은 금방 할 것 같습니다.

이외에도 사용되지만 언급하지 않은 파라미터들 수치에 대한 제안, data 불러오는 파트에서 추가적인 최적화,
또는 그 이외에 여러가지 의견을 주시면 감사하겠습니다.
이 코드는 실제로 돌려서 학습이 가능한 것을 확인하고 올립니다.

### CODE REVIEW

File: config_sampler_test.py [hunk 1]
```diff
@@ -35,9 +35,26 @@ def test_conv_temporal_sampler(self):
...
+            'linear': {'activation': ['relu', None], 'units': [8, 16]},
+            'gru': {'units': [39]}
+        }
+        default_config = {'last_units': 1}
+        model_config = vad_architecture_sampler(search_space_2d,
+                                                search_space_1d,
+                                                n_blocks=4,
+                                                input_shape=[7, 80, 1],
+                                                default_config=default_config,
+                                                constraint=None)
```
Comment (lsg1213) [inline config_sampler_test.py:57]: 이건 그냥 궁금해서 여쭤보는 건데 테스트 도중에 터지는지 여부만 확인하는 건가요?

File: vad_dataloader.py [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+    label_fnames = [os.path.join(label_folder, 
+                                 os.path.split(fname)[1].replace('wav', 'npy'))
+                    for fname in wav_fnames]
+    return wav_fnames, label_fnames
+
+
+def extract_feat_label(wav_fname, label_fname, **kwargs):
+    wav = load_wav(wav_fname, **kwargs)
+    label = load_label(label_fname, **kwargs)
+    assert len(wav) == len(label)
```
Comment (lsg1213) [inline vad_dataloader.py]: 이건 그냥 개인적인 생각인데 assert에 각각 메시지를 달아주면 나중에 어느 부분에서 터지고 뭐가 문제인지 파악하기 좋지 않을까요?

File: vad_dataloader.py [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+            tf.TensorSpec(shape=(None, n_mels, n_chan), 
+                          dtype=tf.float32),
+            tf.TensorSpec(shape=(None,), dtype=tf.float32)))
+    dataset = dataset.cache()
+    dataset = dataset.map(apply_window(window))
+
+    return dataset
+
+
+def load_wav(name, n_fft=1024, n_mels=80,
```
Comment (lsg1213) [inline vad_dataloader.py]: 함수 이름이 load_wav라고 되어 있는데 load wav뿐만 아니라 변환까지 모두 담고 있어서 load_data 같은 범용적인 이름을 지으면 어떨까요?

File: vad_dataloader.py [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+    if mel_scale is None:
+        mel_scale = get_mel_scale(n_fft, n_mels, sr)
+
+    spec = tf.signal.stft(wav, frame_length=n_fft, frame_step=n_fft//2)
+    spec = tf.abs(spec)
+    spec = tf.matmul(spec, mel_scale)
+    spec = tf.transpose(spec, (1, 2, 0))
+    
+    if logmel:
+        spec = tf.math.log(tf.clip_by_value(spec, 1e-8, tf.reduce_max(spec)))
```
Comment (lsg1213) [inline vad_dataloader.py:89]: 제가 전에 본 자료에서는 log-mel이 상용로그를 쓰는 걸로 알고 있는데 혹시 원래 log-mel이 자연로그를 쓰나요?

File: vad_dataloader.py [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+        spec = (spec - min_value) / (max_value - min_value)
+    return spec
+
+
+def load_label(name, n_fft=1024, **kwargs):
+    label = tf.convert_to_tensor(np.load(name), dtype=tf.float32)
+    label = tf.nn.avg_pool1d(tf.reshape(label, (1, -1, 1)),
+                             n_fft,
+                             n_fft//2, 
+                             padding='VALID')
```
Comment (lsg1213) [inline vad_dataloader.py]: 혹시 tf.frame을 쓰면 좀 더 편하지 않을까 생각합니다

File: config_sampler_test.py (hunk 1 above)
Comment (daniel03c1) [reply config_sampler_test.py:57]: 넵 더 자세하게 돌려볼 수도 있는데, 일단은 그정도만 해서 작성하였습니다.
혹시 더 추가로 테스트할만한 좋은게 있을까요?

File: vad_dataloader.py (hunk 2 above)
Comment (daniel03c1) [reply vad_dataloader.py]: 아 터졌을 때 어디에서 터졌는지 위치는 파악할 수는 있는데, 메세지 다는 건 좋은 것 같습니다.
vad_dataloader에 있는 모든 assert 문에 다 넣을까요?

File: vad_dataloader.py (hunk 3 above)
Comment (daniel03c1) [reply vad_dataloader.py]: wav만 불러와서 전처리하는 코드인데, load_data는 wav 뿐만 아니라 다른 것도 불러오는 것처럼 보여서
조금 헷깔리지는 않을까요?
범용적이라하신 이유는 wav를 불러오기만 하는게 아니라 전처리까지 해서 좀더 이름을 확장시키는건 어떤가
하는 말씀이신거죠? 그러면 load_wav_and_extract 이런식으로 좀 더 자세하게 이름을 바꿀까요?

File: vad_dataloader.py (hunk 4 above)
Comment (daniel03c1) [reply vad_dataloader.py:89]: 아 torchaudio도 그렇고 amplitude를 dB로 바꿀 때는 상용로그 쓰는건 맞는 것 같습니다.
하지만 상용로그나 자연로그나 서로 변환할 때 상수만 곱하거나 나누면 되서, normalize할 때
둘다 차이가 없어져서, 그렇기도 하고, 제가 참고한 태수씨 코드도 이렇게 자연로그로 하고 0-1로
표준화해서 이렇게 진행하였습니다!

File: vad_dataloader.py (hunk 5 above)
Comment (daniel03c1) [reply vad_dataloader.py]: avg_pool1d 대신 tf.signal.frame 사용하고 reduce_mean 사용하자는 말씀이시죠?
좋은 것 같습니다. 한번 반영해보겠습니다.

Comment (daniel03c1) [reply vad_dataloader.py]: 반영하였습니다!

File: vad_dataloader.py [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+    spec = tf.transpose(spec, (1, 2, 0))
+    
+    if logmel:
+        spec = tf.math.log(tf.clip_by_value(spec, 1e-8, tf.reduce_max(spec)))
+
+    if normalize:
+        min_value = tf.reduce_min(spec)
+        max_value = tf.reduce_max(spec)
+        spec = (spec - min_value) / (max_value - min_value)
+    return spec
```
Comment (ParkJinHyeock) [inline vad_dataloader.py:95]: 이렇게 하면 모든 주파수별로 같은 scaling이 되지 않나요? DCASE의 경우 이렇게 되면 문제가 있었는데 이 경우는 log도 씌우고 했으니 괜찮을까요?

File: vad_dataloader.py (hunk 3 above)
Comment (ParkJinHyeock) [reply vad_dataloader.py]: 음 저도 바꿀수있으면 load_wav_and_extract가 더 직관적인것 같습니다. 전처리 하는 플로우 찾는게 약간 헷갈리더라구요

File: vad_dataloader.py (hunk 2 above)
Comment (ParkJinHyeock) [reply vad_dataloader.py]: 사실 밑에 부분이랑 연결이 되는 부분인 output이 wav가 아니라 후처리된 부분인데 wav라고 적혀있어서 조금 더 헷갈렸던 것 같습니다.!

File: vad_dataloader.py [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+    feats_labels = Parallel(n_jobs=8)(
+        delayed(extract_feat_label)(w, l, mel_scale=mel_scale)
+        for w, l in tqdm.tqdm(zip(wav_fnames, label_fnames)))
+
+    joblib.dump(feats_labels, 'libri_aurora_test.jl')
+    '''
+
+    wav_fnames, label_fnames = extract_vad_fnames(WAV_PATH, LABEL_PATH)
+    window = [-19, -10, -1, 0, 1, 10, 19]
+    dataset = get_vad_dataset(wav_fnames, label_fnames, window, n_fft=1024)
```
Comment (ParkJinHyeock) [inline vad_dataloader.py:187]: 혹시 여기서 window 부분은 offset인가요..?

File: models.py [hunk 8]
```diff
@@ -76,3 +77,25 @@ def conv_temporal(input_shape, model_config):
...
+    blocks = sorted([key for key in model_config.keys()
+                     if key.startswith('BLOCK') and not key.endswith('_ARGS')])
+
+    for block in blocks:
+        x = getattr(modules, model_config[block])(
+            model_config[f'{block}_ARGS'])(x)
+
+    x = layers.force_1d_inputs()(x)
+    x = Dense(last_unit, activation='sigmoid')(x)
+    return tf.keras.Model(inputs=inputs, outputs=x)
```
Comment (ParkJinHyeock) [inline models.py:100]: last unit을 변수로 둔 것은 유연성을 위해서 놔두신거죠?

File: vad_train.py [hunk 9]
```diff
@@ -0,0 +1,224 @@
...
+        {'depth': [1, 2, 3],
+         'n_head': [1, 2, 4, 8, 16],
+         'key_dim': [2, 3, 4, 6, 8, 12, 16, 24, 32, 48],
+         'ff_multiplier': [0.25, 0.5, 1, 2, 4, 8],
+         'kernel_size': [1, 3, 5]},
+    'simple_dense_stage':
+        {'depth': [1, 2, 3],
+         'units': [4, 6, 8, 12, 16, 24, 32, 48, 64, 96, 128, 192, 256],
+         'dense_activation': [None, 'relu'],
+         'dropout_rate': [0., 0.2, 0.5]},
```
Comment (ParkJinHyeock) [inline vad_train.py:86]: Dropout rate를 Dense stage에서만 search space에 넣으신 이유가 있을까요? dropout rate은 모델 전체블럭에 비슷하게 넣는건 어떻게 생각하시나요?

File: vad_dataloader.py (hunk 2 above)
Comment (daniel03c1) [reply vad_dataloader.py]: assert에 feat, label 길이가 다르다고 표기하도록 반영하였습니다!

File: vad_dataloader.py (hunk 3 above)
Comment (daniel03c1) [reply vad_dataloader.py]: 네 이름을 load_wav_and_extract_feat로 바로 알 수 있도록 바꾸었습니다!

File: vad_dataloader.py (hunk 6 above)
Comment (daniel03c1) [reply vad_dataloader.py:95]: 모든 주파수별로 같은 scaling이 적용되는 것 맞습니다. DCASE에서 하고 안하고랑 성능 차이가 많이 났나요?
제가 참고한 태수씨 코드도 이렇게 진행해서 이렇게하기도 하였고, 지금까지 하면서 모든 주파수별로 같은 scaling을 적용했을 때
성능이 많이 떨어진적이 없어서 챌린지때 한 것처럼 이렇게 진행했습니다.
혹시 주파수별로 다른 scaling을 적용하는 것, 그리고 같은 scaling을 진행하는 것 두가지의 성능차이가 뚜렷하다면
참고할 자료 공유해주시면 적극 반영해보겠습니다!

File: vad_dataloader.py (hunk 7 above)
Comment (daniel03c1) [reply vad_dataloader.py:187]: 아 여기서는 조금 설명이 필요할 것 같은데, 보통 윈도우하면 [t-5, t-4, t-3, t-2, t-1, t, t+1, t+2, t+3, t+4, t+5] 이렇게 연속적인
윈도우를 사용할 수도 있는데, VAD 논문들 중에서는 저렇게 윈도우를 잡고 작성하는 논문들이 있습니다.
ACAM 등이 여기에 속하는데, 여기에 맞추려고 일단 저렇게 맞춰놓았는데, 윈도우 자체에 대해서도 search space 정할때
정해야하는 파트에 속하긴 합니다. 혹시 진혁씨는 윈도우에 대해서 좋은 아이디어 있으신가요?

File: models.py (hunk 8 above)
Comment (daniel03c1) [reply models.py:100]: last units은 사실 baseline이라고 할 수 있는 DNN, bDNN 모델도 구현할수 있도록 하기 위해서 만들어져있습니다.
bDNN, DNN 모델 모두 [win_size, bins]를 flatten하게 만들어서 [win_size * bins]로 만든 이후에 여기에 dense를 이어붙여
만듭니다. 그래서 최종 unit을 1로 하면 가운데 프레임만 예측하는 DNN이 되고, unit을 win_size로 하게 되면
각 프레임마다 예측하도록 하게 됩니다. 이 구조도 구현할 수 있도록 하기 위해 넣어두었습니다.
사실 frame별로 출력을 넣고, bDNN처럼 flatten하지 않고, 프레임 정보는 건들지 않는 방향으로 간다면, 지워도 무방할 것 같습니다.

File: vad_train.py (hunk 9 above)
Comment (daniel03c1) [reply vad_train.py:86]: 여기에 dropout rate을 사실 굳이 넣지 않으려고 했는데, 넣은 이유는 baseline 모델인 DNN, bDNN에서 
dropout_rate을 0.5로 세팅하고 있기 때문입니다. 저는 사실 빼도 괜찮아보이는데, 진혁씨는 넣을 수 있는 스테이지에는
전부 넣는게 좋아보이시나요?

File: models.py (hunk 8 above)
Comment (ParkJinHyeock) [reply models.py:100]: 아 그렇다면 놔두는게 좋을 것 같습니다!

File: vad_dataloader.py (hunk 6 above)
Comment (ParkJinHyeock) [reply vad_dataloader.py:95]: 그 승진님이 처음에 F가 0~2가 나왔다고 했던게 아마 저런 방식으로 스케일링해서였을텐데, 사실 그 때 더 큰 문제는 train이랑 val이 다른 스케일러를 썼기 때문이여서 정확한 비교는 힘들 것 같네요.. 어차피 지금 성능이 너무 잘나오는 상황이니  상관없을 것 같습니다.

File: vad_train.py (hunk 9 above)
Comment (ParkJinHyeock) [reply vad_train.py:86]: 아 베이스라인 모델이 그렇다면 괜찮을 것 같습니다. 만약 DCASE나 다른 Search space에서 Dropout을 사용한다면 부분적인 Block에만 Dropout을 사용하는 것보다 전체 블럭의 Dropout을 비슷하게 가야 Dropout의 영향력을 제대로 추론이 될 것 같아서 말씀드린거였습니다. 일단 이 Case에서는 놔두는게 좋을 것 같습니다.

**Review by ParkJinHyeock — APPROVED** [review]

File: vad_train.py (hunk 9 above)
Comment (daniel03c1) [reply vad_train.py:86]: 네 말씀하신대로 DCASE나 이런 경우에서 서치에 들어갈 때 dropout도 조절하고 싶다면 말씀하신대로 
dropout이 들어간 모든 stage에도 동일하게 dropout_rate도 넣는것이 좋을 것 같습니다.

File: config_sampler_test.py (hunk 1 above)
Comment (lsg1213) [reply config_sampler_test.py:57]: 음.. 딱히 떠오르는 건 없습니다

**Review by lsg1213 — APPROVED** [review]
